pkg/agents/team4/finalAgent/agent.go

```diff
@@ -36,153 +49,202 @@ type CustomAgentEvo struct {
+
+	//call the relevant message handler
+	a.callHandleMessage()
+
+	a.sendingMessage()
+
+	a.checkAgentResponseMemory() // check if we received the Hp message and intended food message and update trust score
+
+	a.setPersonality()
+
+	// if platform is on floor or floor below match agent response to sent message
+	if a.PlatformOnFloor() || (!a.PlatformOnFloor() && a.CurrPlatFood() != -1) {
+		a.verifyResponses()
 	}
 
-	foodEaten := food.FoodType(0)
+	//----------------------EATING FOOD -------------------------//
+
+	var foodEaten food.FoodType
 	var err error
-	calculatedAmountToEat := food.FoodType(0)
 
-	if (a.Age()-a.params.ageLastEaten) >= a.params.daysToWait[a.params.currentPersonality][a.params.healthStatus] || a.params.healthStatus == 0 {
-		a.params.locked = false
-		calculatedAmountToEat = food.FoodType(a.params.traumaScaleFactor * float64(a.params.foodToEat[a.params.currentPersonality][a.params.healthStatus]))
-		foodEaten, err = a.TakeFood(food.FoodType(calculatedAmountToEat))
-		a.params.ageLastEaten = a.Age()
+	if rand.Intn(100) >= (a.params.waitProbability[a.params.currentPersonality][a.params.healthStatus]-a.params.craving) && !a.HasEaten() && a.PlatformOnFloor() {
+		a.params.intendedFoodToTake = food.FoodType(a.params.foodToEat[a.params.currentPersonality][a.params.healthStatus])
+
+		// prioritises treaties over our current food behaviour
+		a.handleActiveTreatyConditions()
+
+		// eat and update last food
+		foodEaten, err = a.TakeFood(a.params.intendedFoodToTake)
+		a.params.lastFoodTaken = foodEaten
+		// records whether there was food taken
+		if foodEaten > 0 {
+			a.params.ageLastEaten = a.Age()
+		}
+
+		// setting craving to 0 if we are fulfilled otherwise reduce the craving proportional to foodEaten
+		if a.params.foodToEat[a.params.currentPersonality][a.params.healthStatus] != 0 {
+			a.params.craving -= a.params.craving * (int(foodEaten) / a.params.foodToEat[a.params.currentPersonality][a.params.healthStatus])
+		} else {
+			a.params.craving = 0
+		}
+
+		// error handling
 		if err != nil {
 			switch err.(type) {
 			case *infra.FloorError:
+				log.Error("Simulation - team4/agentTraining.go: \t FloorError: is the platform on your floor?")
 			case *infra.NegFoodError:
+				log.Error("Simulation - team4/agentTraining.go: \t NegFoodError: is calculatedAmountToEat negative?")
 			case *infra.AlreadyEatenError:
+				log.Error("Simulation - team4/agentTraining.go: \t AlreadyEatenError: Have you already eaten?")
 			default:
 			}
 		}
 	}
 
-	a.Log("team4EvoAgent reporting status:", infra.Fields{"floor": a.Floor(), "hp": a.HP(), "FoodToEat": calculatedAmountToEat, "DaysToWait": a.params.daysToWait, "foodEaten": foodEaten})
+	//log agent state
+	a.Log("team4EvoAgent reporting status:", infra.Fields{"floor": a.Floor(), "hp": a.HP(), "FoodToEat": a.params.intendedFoodToTake, "WaitProbability": a.params.waitProbability, "foodEaten": foodEaten})
 }
```

Comment:
Please break this up into functions for readability:

- You can put everything inside the `if a.hasReshuffled()` statement into a function
- The second if statement should be in a function too